eng/targets/CheckLongPathSupport.targets

```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  'path-too-long'-errors may occur very early when building / loading projects, so we ensure this check get's executed by utilizing 'InitialTargets',
+  see https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-specify-which-target-to-build-first?view=vs-2022#use-the-initialtargets-attribute
+-->
+<Project InitialTargets="EnsureLongPathSupport">
```

Comment:
Using a `<Project>` here seems a bit heavy. Why are you not just replicating the target that Roslyn is using 

```xml
  <Target Name="_CheckLongPathSupport" BeforeTargets="BeforeBuild" Condition="'$(MSBuildRuntimeType)' == 'Full'">
    <PropertyGroup>
      <_RoslynLongPathsEnabled>$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control\FileSystem', 'LongPathsEnabled', null, RegistryView.Registry64, RegistryView.Registry32))</_RoslynLongPathsEnabled>
    </PropertyGroup>
    
    <Warning Condition="'$(_RoslynLongPathsEnabled)' != '1'" Text="Long paths are required for this project. Please run eng\enable-long-paths.reg" />
  </Target> 
```

Reply:
Originally intended to do so. However, when relying on `BeforeTargets="BeforeBuild"` this target won't execute. Instead, only this generic 

`error MSB4248: Cannot expand metadata in expression "$([MSBuild]::ValueOrDefault('%(FullPath
)', '').StartsWith($([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))))". The item metadata "%(FullPath)" ca
nnot be applied to the path "TestFiles\IntegrationTests\ComponentDesignTimeCodeGenerationTest\GenericComponent_GenericE
ventCallbackWithGenericTypeParameter_NestedTypeInference\TestComponent.mappings.txt".`

Error pops up during `build.cmd`.

I guess this is due to the fact that above error occurs too early (at project load time) and therfore `BeforeBuild` never gets executed. see previous [comment](https://github.com/dotnet/razor/pull/11140#issuecomment-2455440939)

As for the different check when to do the registry-lookup:

I felt it would be a bit cleaner to actually check for the paltform (as of https://github.com/dotnet/msbuild/issues/539)  instead of relying on some implicit side-effect via (`Condition="'$(MSBuildRuntimeType)' == 'Full'"`).

Wouldn't fight about this, though ;)

Reply:
Questinable might be: Should this really be done via `msbuild`-mechanisms, or might it be sufficient to just include it in `build.-cmd`, which is windows-only by design ;).

I preferred the first, because then this warning would also be generated when doing an ordinary `dotnet build` - not sure if this is actually supported / recommended, though.

Reply:
> Originally intended to do so. However, when relying on BeforeTargets="BeforeBuild" this target won't execute. Instead, only this generic

Do we have a binlog where this error occurs? That seems very off. 

Basically my mental model is that this works for roslyn, why can't it work for razor? Nested project builds are also one of my "that is probably going to have impact we don't like so be wary". This is why I'm focused on this part of the change.

Reply:
@jaredpar Gathering a `binlog` is new to me, but seems to be straight forward to collect. I will have a look into this. Very appreciating  someone paying attention so much attention into such nitty-gritty details 👌.
Will keep you updated.

Reply:
@jaredpar 
Doing it the "roslyn" way produces this binlog:

![image](https://github.com/user-attachments/assets/5f653a64-3287-43e2-ac39-8d612ebb99d9)

> Cannot expand metadata in expression "$([MSBuild]::ValueOrDefault('%(FullPath)', '').StartsWith($([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))))". The item metadata "%(FullPath)" cannot be applied to the path "TestFiles\IntegrationTests\ComponentDesignTimeCodeGenerationTest\GenericComponent_GenericEventCallbackWithGenericTypeParameter_NestedTypeInference\TestComponent.mappings.txt". Path: {{PATH_TO_REPOSITORY}}\razor\src\Compiler\Microsoft.AspNetCore.Razor.Language\test\TestFiles\IntegrationTests\ComponentDesignTimeCodeGenerationTest\GenericComponent_GenericEventCallbackWithGenericTypeParameter_NestedTypeInference\TestComponent.mappings.txt exceeds the OS max path limit. The fully qualified file name must be less than 260 characters.

The error seems to occur in the `_GetAllRestoreProjectPathItems`-Target, which underpins my observation of `BeforeBuild` is never reached in this scenario.

If we might identify another Target which we could youse to "hook into", this would be totally fine. But [from the docs](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-targets?view=vs-2022#sdk-and-default-build-targets) I couldn't identify a good candidate - hence the approach utilizing `InitialTarget`.

Reply:
@baronfel, @rainersigwald why would we be getting this error here but not roslyn?

Reply:
@jaredpar just fyi:
Roslyn may suffer from the same error when the base path is too long for certain projects. However, the at least it displays the "long-path-warning" for all others.

Warnings (from target):
![image](https://github.com/user-attachments/assets/c214cd1b-891a-4c22-a39b-6a636d2e6700)

Error due to too long path (like here):
![image](https://github.com/user-attachments/assets/62abc8df-b2ed-43d1-bb2e-6a5b391c60f4)

Reply:
`BeforeTargets="BeforeBuild"` is very early in the "real" build process, and is a great place to hang a hook like this. However, it doesn't account for all types of builds.

`InitialTargets` _does_ work for all types of builds, so it'll fire in the restore scenario that you're seeing here. However, it also fires in design-time builds, where you don't really want to spend the time doing this check, as they're pretty latency-sensitive.

You could maybe do `BeforeTargets="BeforeBuild;CollectPackageReferences"`?

Reply:
@rainersigwald thx, this did the trick. Exactly an alternative I was hoping for 🚀.
@jaredpar Check is now in line with what roslyn does (but it comes with a handy link to the docs within VS):
![image](https://github.com/user-attachments/assets/070af533-4085-4866-b9c5-ebd1bdb41e05)

Or would it be a better practice to have this ErrorCode be something like `RZ1337` ? But this might be confusing when it comes to **real** diagnostics, right?